packages/typespec-client-generator-core/test/decorators.test.ts

```diff
@@ -3027,4 +3027,182 @@ describe("typespec-client-generator-core: decorators", () => {
+      const runnerWithCSharp = await createSdkTestRunner({
+        emitterName: "@azure-tools/typespec-csharp",
+      });
+      const diagnostics = await runnerWithCSharp.diagnose(`
+        @service
+        @test namespace MyService {
+          @test
+          @clientName("TestRenamed", "csharp, !java")
+          model Test {
+            prop: string;
+          }
+          @test
+          @access(Access.internal)
+          op func(
+            @body body: Test
+          ): void;
+        }
+      `);
+
+      expectDiagnostics(diagnostics, {
+        code: "@azure-tools/typespec-client-generator-core/invalid-negation-scope",
+      });
+    });
+
+    it("disallow combination of negation scope and normal scope, starting with !", async () => {
+      const runnerWithCSharp = await createSdkTestRunner({
+        emitterName: "@azure-tools/typespec-csharp",
+      });
+      const diagnostics = await runnerWithCSharp.diagnose(`
+        @service
+        @test namespace MyService {
+          @test
+          @clientName("TestRenamed", "!csharp, java")
+          model Test {
+            prop: string;
+          }
+          @test
+          @access(Access.internal)
+          op func(
+            @body body: Test
+          ): void;
+        }
+      `);
+
+      expectDiagnostics(diagnostics, {
+        code: "@azure-tools/typespec-client-generator-core/invalid-negation-scope",
+      });
+    });
+
+    it("allow multiple separated negation scopes", async () => {
+      const runnerWithCSharp = await createSdkTestRunner({
+        emitterName: "@azure-tools/typespec-csharp",
+      });
+      await runnerWithCSharp.diagnose(`
+        @service
+        @test namespace MyService {
+          @test
+          @clientName("TestRenamed", "!csharp, !java")
+          model Test {
+            prop: string;
+          }
+          @test
+          @access(Access.internal)
+          op func(
+            @body body: Test
+          ): void;
+        }
+      `);
+
+      const sdkPackage = runnerWithCSharp.context.sdkPackage;
+      const testModel = sdkPackage.models.find((x) => x.name === "Test");
+      ok(testModel);
+    });
+
+    it("negation scope override normal scope", async () => {
+      const runnerWithCSharp = await createSdkTestRunner({
+        emitterName: "@azure-tools/typespec-csharp",
+      });
+      await runnerWithCSharp.diagnose(`
+        @service
+        @test namespace MyService {
+          @test
+          @clientName("TestRenamedAgain", "!csharp, !java")
+          @clientName("TestRenamed", "csharp")
+          model Test {
+            prop: string;
+          }
+          @test
+          @access(Access.internal)
+          op func(
+            @body body: Test
+          ): void;
+        }
+      `);
+
+      const sdkPackage = runnerWithCSharp.context.sdkPackage;
+      const testModel = sdkPackage.models.find((x) => x.name === "Test");
```

Comment:
why this name is `Test`? i suppose it should be `TestRenamed`?

Reply:
```
          @clientName("TestRenamedAgain", "!csharp, !java")
          @clientName("TestRenamed", "csharp")
```
Since we define that later decorator overrides the former decorator, `@clientName("TestRenamedAgain", "!csharp, !java")` will take effect, right? Then this renaming will exclude csharp, so the model remains "Test".

Reply:
the override is by single scope which mean you could set value for java, python, go with three separated decorators. then the problem is how should we handle the negation part.

Reply:
```
@clientName("A", "csharp")
@clientName("B", "java")
Model Foo
```
For normal scopes, we can just add them incrementally.

```
@clientName("A", "!csharp")
@clientName("B", "!java")
Model Foo
```
For the negation scope override, the value of other scopes changes. We can't just add the negated scopes to the original map.
Two options here:
- Simply override everything when a negation scope shows up including the others' value and the negation scopes
- Throw error if others' values are different with different negation scopes